Rework SIFA domains

schuessf opened the pull request:
* Implement octagon domain in SIFA
* Implement equality domain in SIFA
* Introduce abstractions with composition for state-based SIFA domains with DNF (for intervals and octagons) and without (for equality domain)
* Add abstraction for non-relational domains, used by intervals for now

schuessf committed:
Move OctMatrix and OctValue to SmtLibUtils

schuessf committed:
Add skeleton for OctagonDomain in SIFA

schuessf committed:
Add settings for OctagonDomain

schuessf committed:
Fix matrix initialization

schuessf committed:
Use OctagonRelation, initial attempt to convert Term to OctMatrix

schuessf committed:
Modify matrix instead of intersecting

schuessf committed:
Move RewriteEqualityTransformer to own class in SmtLibUtils

schuessf committed:
Rewrite equalities before DNF

schuessf committed:
Fix relation normalization, minor changes

schuessf committed:
Add widening operator

schuessf committed:
Implement join and widen similar to intervals

schuessf committed:
Optimization for false

schuessf committed:
Add settings for regression tests

schuessf committed:
Move octagon settings to subfolders

schuessf committed:
Rearrange matrices before merge

schuessf committed:
Add toString method

schuessf committed:
Use abstract class AbstractStateBasedDomain for octagons

schuessf committed:
Use AbstractStateBasedDomain for IntervalDomain

schuessf committed:
Move main code for toTerm to Interval, CC fixes

schuessf committed:
Allow to transform the term before the DNF
and rewrite equalities in OctagonDomain

schuessf committed:
Minor simplification

schuessf committed:
Reduce visibility

schuessf committed:
Use timer in IntervalDomain

schuessf committed:
Add headers and documentation

schuessf committed:
Bugfix: Missing ! because of different representations

schuessf committed:
Minor: Use SmtUtils methods in OctMatrix::getTerm

schuessf committed:
Avoid unnecessary matrix rearrangement, minor fix

schuessf committed:
Minor: Rename fields

schuessf committed:
Introduce super class and interface for nonrelational domains

danieldietsch commented on trunk/source/AbstractInterpretationV2/src/de/uni_freiburg/informatik/ultimate/plugins/analysis/abstractinterpretationv2/domain/relational/octagon/OctInterval.java:
```diff
@@ -67,11 +70,17 @@ public OctInterval(final IntervalDomainValue ivlInterval) {
 		}
 	}
 
+	private static OctValue toOctValue(final IntervalValue interval) {
```
👍

danieldietsch commented on trunk/source/AbstractInterpretationV2/src/de/uni_freiburg/informatik/ultimate/plugins/analysis/abstractinterpretationv2/domain/relational/octagon/OctInterval.java:
```diff
@@ -128,7 +137,11 @@ public IntervalDomainValue toIvlInterval() {
+	}
+
+	private static IntervalValue toIntervalValue(final OctValue octValue) {
+		return octValue.isInfinity() ? new IntervalValue() : new IntervalValue(octValue.getValue());
```
Perhaps also use a static and pre-created top IntervalValue ?

schuessf replied:
Sure, I can do that, but this is not really related to SIFA. I just had to replace the constructor `OctValue(IntervalValue)` with this static method, because these classes are now in different plugins.

schuessf committed:
Rename setting correctly

schuessf committed:
Minor: Sort vars by name

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/AbstractStateBasedDomain.java:
```diff
@@ -0,0 +1,170 @@
+	protected final SymbolicTools mTools;
+	private final int mMaxDisjuncts;
+	protected final Supplier<IProgressAwareTimer> mTimeout;
+	private final WeakHashMap<IPredicate, List<STATE>> mPredicateCache = new WeakHashMap<>();
```
Is a weak hashmap the right choice here? If I remember correctly, the states are thrown away relatively soon, so it might even slow down everything.

schuessf replied:
I don't know. It was previously used in the `IntervalDomain`, so I simply used the same here.

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/AbstractStateBasedDomain.java:
```diff
@@ -0,0 +1,170 @@
+			}
+			final STATE state = toState(SmtUtils.getConjuncts(dnfDisjunct));
+			if (state != null) {
+				result.add(state);
```
why should the return be null?

schuessf replied:
The null should represent a bottom-state here, but this is not a good solution! I think the best way is to add a `isBottom` method to `IAbstractState`.

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/AbstractStateBasedDomain.java:
```diff
@@ -0,0 +1,170 @@
+	}
+
+	private IPredicate toPredicate(final Collection<STATE> states) {
+		return mTools.orT(states.stream().map(x -> x.toTerm(mTools.getScript())).collect(Collectors.toList()));
```
I dont like the name `orT` -- and it looks like its never documented. What is the difference between `or` and `orT`? Same for `and` and `andT`

schuessf replied:
Looks like the "T" stands for term, therefore the `or`-method takes a collection of `IPredicate`, whereas the `orT`-method takes a collection of `Term`. But we can also simply overload these methods 😉

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/IAbstractState.java:
```diff
@@ -0,0 +1,44 @@
+ * @param <STATE>
+ *            The type of the state
+ */
+public interface IAbstractState<STATE extends IAbstractState<STATE>> extends ITermProviderOnDemand {
```
It might not be clear to an implementer what the difference between join and widen is. Can you add comments here?

schuessf replied:
👍

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/INonrelationalValue.java:
```diff
@@ -0,0 +1,47 @@
+
+	VALUE union(VALUE other);
+
+	VALUE widen(VALUE other);
```
Add comments to the interface methods (in particular union vs widen).

schuessf replied:
👍

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/IntervalDomain.java:
```diff
@@ -209,81 +76,89 @@ private Optional<Map<TermVariable, Interval>> dnfDisjunctToIntervals(
 					continue;
 				} else if (updatedLhsInterval.get().isBottom()) {
-					return Optional.empty();
+					return null;
```
Shouldnt that return BOT?

schuessf replied:
Yes, I will change that.

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/IntervalDomain.java:
```diff
@@ -209,81 +76,89 @@ private Optional<Map<TermVariable, Interval>> dnfDisjunctToIntervals(
-		return solvedRelations;
+	@Override
+	protected NonrelationalState<Interval> getTopState() {
+		return new NonrelationalState<>();
```
Shouldnt that return the unified top state?

schuessf replied:
👍

schuessf replied:
Okay, this is more complicated, since NonrelationalState is parametrized. I would just leave it like that (it is just called in case of a timeout anyway), instead of doing some unsafe casts 😉

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/IntervalDomain.java:
```diff
@@ -209,81 +76,89 @@ private Optional<Map<TermVariable, Interval>> dnfDisjunctToIntervals(
+				}
 			}
 		}
+		Collections.sort(result, new CompareNumberOfFreeVariablesInRhs(result));
```
why is it important to sort the result?

schuessf replied:
Not sure, this was also done before the PR (in `dnfDisjunctToIntervals`), I just had to refactor this to deal with an array instead of a single term.

danieldietsch commented on trunk/source/Library-Sifa/src/de/uni_freiburg/informatik/ultimate/lib/sifa/domain/OctagonDomain.java:
```diff
@@ -0,0 +1,71 @@
+	@Override
+	protected Term transformTerm(final Term term) {
+		// TODO consider removing boolean sub-terms before computing DNF as we don't use the boolean terms anyways
+		return new RewriteEqualityTransformer(mTools.getScript()).transform(term);
```
Would it make sense to keep an instance of RewriteEqualityTransformer around and reuse that instead of recreating for every call?

schuessf replied:
👍

schuessf committed:
Return bottom state instead of null

schuessf committed:
Store an instance of the term transformer

schuessf committed:
Add toString

schuessf committed:
Minor: Rename method for consistency

schuessf committed:
Add some documentation

schuessf committed:
Add TODO

schuessf committed:
Update benchexec file to evaluate Interval/Octagon

schuessf committed:
Add missing propertyfile

danieldietsch committed:
fix paths

danieldietsch committed:
fix set files, fix more paths

schuessf committed:
Add settings for EqDomain

schuessf committed:
Use SmtUtils.distinct to respect normal form

schuessf committed:
Allow > and < to be parsed by EqDomain

schuessf committed:
Implement EqDomain

schuessf committed:
Introduce AbstractDnfStateBasedDomain as subclass of AbstractStateBasedDomain

schuessf committed:
Rename abstract classes

schuessf committed:
Add documentation

schuessf committed:
Add equality settings for regression tests

schuessf committed:
Bugfix: Freeze state after merge

schuessf committed:
Overapproximate quantifiers instead of crashing

schuessf committed:
Add widening operator that always return top

schuessf committed:
Bugfix: Use the right rounding method

schuessf committed:
Use composition instead of abstract classes

schuessf committed:
Add EqDomain to benchexec file

schuessf committed:
Let octagons in SIFA use their own matrix
This matrix has only the methods that are needed for SIFA and it uses Rational instead of BigDecimal (that are wrapped in OctValue).

schuessf committed:
Revert "Move OctMatrix and OctValue to SmtLibUtils"

schuessf committed:
Bugfix: Ignore quantified formulas

danieldietsch committed:
fix space

schuessf committed:
Use constructor instead of static method

schuessf committed:
Minor

schuessf committed:
Minor: Avoid unnecessary list
